source/common/http/http1/codec_impl.cc

```diff
@@ -1323,11 +1369,66 @@ void ServerConnectionImpl::ActiveRequest::dumpState(std::ostream& os, int indent
   os << DUMP_MEMBER(response_encoder_.local_end_stream_);
 }
 
+Http::HeaderValidatorPtr ServerConnectionImpl::makeHeaderValidator() {
+  Http::HeaderValidatorPtr validator;
+#ifdef ENVOY_ENABLE_UHV
+  if (header_validator_factory_) {
+    validator = header_validator_factory_->create(Http::HeaderValidatorFactory::Protocol::HTTP1,
+                                                  active_request_->request_decoder_->streamInfo());
+  }
+#endif
+  return validator;
+}
+
+Http::Status ServerConnectionImpl::checkHeaderEntry([[maybe_unused]] const HeaderString& key,
+                                                    [[maybe_unused]] const HeaderString& value) {
+  Http::Status result;
+#ifdef ENVOY_ENABLE_UHV
+  if (active_request_->header_validator_) {
+    auto validation_result =
+        active_request_->header_validator_->validateRequestHeaderEntry(key, value);
+    if (validation_result == Http::HeaderValidator::HeaderEntryValidationResult::Reject) {
+      result = codecProtocolError(
+          absl::StrCat("http/1.1 protocol error: invalid header ", key.getStringView()));
+    }
+  }
+#endif
+  return result;
+}
+
+Http::Status
+ServerConnectionImpl::checkHeaderMapAndHandleError([[maybe_unused]] RequestHeaderMap& header_map) {
+  Http::Status result;
+#ifdef ENVOY_ENABLE_UHV
+  if (active_request_->header_validator_) {
+    auto validation_result =
+        active_request_->header_validator_->validateRequestHeaderMap(header_map);
+    if (validation_result == Http::HeaderValidator::RequestHeaderMapValidationResult::Reject ||
+        validation_result == Http::HeaderValidator::RequestHeaderMapValidationResult::Redirect) {
+      result = codecProtocolError("http/1.1 protocol error: invalid header map");
+      std::function<void(ResponseHeaderMap & headers)> modify_headers;
+      if (validation_result == Http::HeaderValidator::RequestHeaderMapValidationResult::Redirect &&
+          !Grpc::Common::hasGrpcContentType(header_map)) {
```

Comment:
Also I _think_ before we mostly did checks of each header after processing all, so if we had a gRPC request and a bad header before gRPC content type, response would follow gRPC path, and now we'll send a standard error reply for headers we would have done "in envoy" (same path for prior http_parser errors).  I don't think it's deal breaker but we should consider documenting this wherever we document differences between the old and the new.

It's also a lot more work to do this in all three codecs.  Given we're probably going to have to have a "check these headers for errors" function for headers transformed by WASM or ext_authz and such, I wonder if we should simply do a one time pass on all headers after we read them and before the HCM processes them, rather than doing all this invasive plumbing into client and server codecs for HTTP/1, HTTP/2 and HTTP/3.  We're going to have to do a pass against all headers once we have all headers to make sure we catch things like 'no content length and transfer encoding' so I wonder if we should just do character validation there and skip all the work along the way.  Thoughts?

Reply:
I have moved header validation into the HCM and moved it out of the codec. I think if we need to do checks for individual headers as they are being parsed, I can plumb it back into codecs. It is however performance optimization only, that allows header parsing to be stopped early once bad header is found.